File: server/app/typing/feeds.py
"""Feed schemas - DRY approach using SQLModel as base."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator

from app.models.enums import FeedCategory
from app.typing.common import response_config
from app.typing.entries import ArticleCreate

# ================= Base (Field Bundles) =================


class FeedBase(BaseModel):
    """Core feed fields - reused across schemas."""

    url: str
    title: str = Field(...)
    description: str = "Follow recent articles from this feed"
    link: str | None = None
    language: str = "en"
    image_url: str | None = None

    @classmethod
    def empty_str_to_none(cls, v):
        return v or None


# ================= Requests =================


class FeedCreate(BaseModel):
    """Input for discovering/adding a feed."""

    url: HttpUrl
    folder_id: UUID | str = "default"


class AdminFeedUpdate(BaseModel):
    """Admin-only schema for updating global feed properties."""

    title: str | None = None
    description: str | None = None
    link: HttpUrl | None = None
    language: str | None = None
    image_url: str | None = None
    url: str | None = None
    top_level_category: FeedCategory | None = None
    popularity_score: float | None = None
    tags: list[str] | None = None


# ================= Responses =================


class FeedSummary(FeedBase):
    """
    Lightweight feed info.
    Used in: Subscriptions List, Search Results, embedded in articles.
    """

    model_config = response_config

    id: UUID


class FeedDetail(FeedSummary):
    """
    Heavy feed info - extends summary with full metadata.
    Used in: Feed Settings, Inspector.
    """

    description: str | None = None
    language: str = "en"

    # Fetching Logic
    last_fetched_at: datetime | None = None
    next_fetch_at: datetime | None = None
    adaptive_fetch_interval_minutes: int | None = None
    last_error_message: str | None = None
    error_count: int = Field(default=0, serialization_alias="error_count")

    # Advanced Metadata
    popularity_score: float = 0.0
    subscriber_count: int = 0

    top_level_category: FeedCategory | None = None
    tags: list[str] = Field(default_factory=list)

    @field_validator("tags", mode="before")
    @classmethod
    def none_to_list(cls, v):
        return v or []

    # Subscription status (for preview mode)
    is_subscribed: bool = False

    # Timestamps
    created_at: datetime
    last_updated_at: datetime | None = None


class FeedEnrichmentResponse(BaseModel):
    """
    AI enrichment response for feed metadata.
    Used in: Batch feed enrichment via Gemini API.
    """

    enhanced_description: str | None = None
    tags: list[str] = Field(default_factory=list)
    category: str
    popularity_estimate: int = Field(ge=1, le=100)


class FeedEnrichmentInput(BaseModel):
    """
    Input for batch feed enrichment.
    """

    title: str
    description: str | None = None
    domain: str
    language: str = "en"
    link: str | None = None
    url: str | None = None
    tags: list[str] = Field(default_factory=list)
    contributors: list[str] = Field(default_factory=list)
    articles: list[str] = Field(default_factory=list)


class FaviconResult(BaseModel):
    """Result from favicon extraction."""

    image_url: str | None = None
    canonical_link: str | None = None


class ArticleStats(BaseModel):
    """Statistics about recent articles for quality scoring."""

    count: int = 0
    image_ratio: float = 0.0
    avg_content_length: float = 0.0
    days_since_last_article: int = 999


class FeedScoringData(BaseModel):
    """Feed metadata used for scoring."""

    id: UUID | str | None = None
    title: str | None = None
    description: str | None = None
    image_url: str | None = None
    language: str | None = None
    domain: str | None = None


class FeedEnrichmentSnapshot(FeedScoringData):
    """Snapshot of feed data for enrichment processing."""

    link: str | None = None
    url: str | None = None
    article_stats: ArticleStats | None = None


class MeilisearchFeedDocument(BaseModel):
    """Feed document structure for Meilisearch indexing."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    url: str
    title: str
    description: str | None = None
    link: str | None = None
    language: str | None = None
    image_url: str | None = None
    tags: list[str] = Field(default_factory=list)
    top_level_category: str | None = None
    popularity_score: float = 0.0

    @field_validator("top_level_category", mode="before")
    @classmethod
    def convert_enum_to_str(cls, v):
        if hasattr(v, "value"):
            return v.value
        return v


class ParsedFeed(BaseModel):
    """
    Schema for feed preview data.
    Returned by the discovery/preview endpoint.
    """

    title: str
    id: str | None = None
    url: str | None = None
    description: str | None = None
    link: str | None = None
    language: str | None = None
    image_url: str | None = None
    last_updated_at: datetime | None = None
    tags: list[str] = Field(default_factory=list)
    articles: list["ArticleCreate"] = Field(default_factory=list)


class BulkDeleteResponse(BaseModel):
    """Response for bulk delete operations."""

    deleted_count: int
    deleted_ids: list[UUID]


class BulkUpdateResponse(BaseModel):
    """Response for bulk update operations."""

    updated_count: int

File: packages/shared/src/api/types/feeds.ts
import { FolderResponse } from './folders';

// Enums from server/app/models/enums.py
export enum FeedCategory {
  ARTS_CULTURE = 'arts_culture',
  AUTOMOTIVE_TRANSPORT = 'automotive_transport',
  BUSINESS_FINANCE = 'business_finance',
  CONSUMER_TECH_DIGITAL = 'consumer_tech_digital',
  ENTERTAINMENT = 'entertainment',
  FAMILY_RELATIONSHIPS = 'family_relationships',
  FOOD_DRINK = 'food_drink',
  GAMING = 'gaming',
  HEALTH_WELLNESS = 'health_wellness',
  HOME_HOBBIES = 'home_hobbies',
  IDENTITY_COMMUNITY = 'identity_community',
  INDUSTRY_PROFESSIONS = 'industry_professions',
  NEWS_CURRENT_EVENTS = 'news_current_events',
  REGIONAL_LOCAL = 'regional_local',
  SCIENCE_NATURE = 'science_nature',
  SOCIETY_LAW_HISTORY = 'society_law_history',
  SOFTWARE_ENGINEERING = 'software_engineering',
  SPORTS = 'sports',
  STYLE_SHOPPING = 'style_shopping',
  TRAVEL_GEOGRAPHY = 'travel_geography',
  MISCELLANEOUS = 'miscellaneous',
}

export enum ContentType {
  AGGREGATOR = 'aggregator',
  CORPORATE_BLOG = 'corporate_blog',
  DOCUMENTATION_WIKI = 'documentation_wiki',
  EDUCATION_RESEARCH = 'education_research',
  FORUM_COMMUNITY = 'forum_community',
  GOVERNMENT_INSTITUTIONAL = 'government_institutional',
  INDIE_BLOG = 'indie_blog',
  MAGAZINE_EDITORIAL = 'magazine_editorial',
  MARKETPLACE_LISTINGS = 'marketplace_listings',
  NEWSLETTER = 'newsletter',
  NEWS_OUTLET = 'news_outlet',
  OPEN_SOURCE_ACTIVITY = 'open_source_activity',
  PODCAST_FEED = 'podcast_feed',
  STATUS_CHANGELOG = 'status_changelog',
  VIDEO_CHANNEL = 'video_channel',
}

// Feed related types

export interface FeedBase {
  url: string;
  title: string;
  description: string | null;
  link: string | null;
  language: string;
  image_url: string | null;
  author: string | null;
  content_type: ContentType | null;
  tags_native: string[];
}

export interface FeedSummary extends FeedBase {
  id: string;
  is_subscribed?: boolean;
  top_level_category?: FeedCategory | null;
  popularity_score?: number;
}

export interface FeedDetail extends FeedSummary {
  // Fetching Logic
  last_fetched_at: string | null;
  next_fetch_at: string | null;
  adaptive_fetch_interval_minutes: number | null;
  last_error_message: string | null;
  error_count: number;

  // Advanced Metadata
  popularity_score: number;
  subscriber_count: number;
  top_level_category: FeedCategory | null;
  tags: string[];

  is_subscribed: boolean;

  // Timestamps
  created_at: string;
  last_updated_at: string | null;
}

// Requests

export interface FeedCreate {
  url: string;
  folder_id?: string;
}

export interface AdminFeedUpdate {
  title?: string;
  description?: string;
  link?: string;
  language?: string;
  image_url?: string;
  url?: string;
  top_level_category?: FeedCategory | null;
  popularity_score?: number;
  tags?: string[];
  tags_native?: string[];
  author?: string;
  content_type?: ContentType | null;
}

// Subscriptions

export interface SubscriptionBase {
  is_favorite: boolean;
  custom_title: string | null;
}

export interface SubscriptionCreate {
  url: string;
  folder_id?: string;
  custom_title?: string;
}

export interface SubscriptionCreateByFeedId {
  folder_id?: string;
  custom_title?: string;
  is_favorite?: boolean;
}

export interface SubscriptionUpdate {
  custom_title?: string;
  folder_id?: string;
  is_favorite?: boolean;
}

export interface SubscriptionResponse extends SubscriptionBase {
  id: string;
  feed: FeedSummary;
  folder: FolderResponse; // NOTE: Backend returns full FolderResponse now
  created_at: string;
}

export interface SubscriptionResponseExtended extends SubscriptionResponse {
  feed: FeedDetail;
}

// Aliases for compatibility
export type Subscription = SubscriptionResponse;
export type SubscriptionExtended = SubscriptionResponseExtended;
export type Feed = FeedSummary; // Or FeedDetail, depending on usage context, but mostly Summary in lists.

// Enrichment & Discovery

export interface FeedEnrichmentResponse {
  enhanced_description: string | null;
  tags: string[];
  tags_native: string[];
  author: string | null;
  content_type: string | null;
  category: string;
  popularity_estimate: number;
}

export interface FeedEnrichmentInput {
  title: string;
  description?: string;
  domain: string;
  language?: string;
  link?: string;
  url?: string;
  tags?: string[];
  contributors?: string[];
  articles?: string[];
}

export interface FaviconResult {
  image_url: string | null;
  canonical_link: string | null;
}

export interface FeedScoringData {
  id?: string;
  title?: string;
  description?: string;
  image_url?: string;
  language?: string;
  domain?: string;
  tags: string[];
  tags_native: string[];
  author?: string;
  content_type?: ContentType | null;
}

export interface MeilisearchFeedDocument extends FeedScoringData {
  id: string;
  url: string;
  link: string | null;
  top_level_category: string | null;
  popularity_score: number;
}

export interface ParsedFeed {
  title: string;
  id: string | null;
  url: string | null;
  description: string | null;
  link: string | null;
  language: string | null;
  image_url: string | null;
  author: string | null;
  content_type: string | null;
  last_updated_at: string | null;
  tags: string[];
  tags_native: string[];
  // articles: ArticleCreate[]; // Avoiding circular dependency for now or import simply
  is_subscribed: boolean;
}

export type FeedDiscoveryResult = ParsedFeed;

export interface FeedsResponse {
  subscriptions: SubscriptionResponse[] | SubscriptionResponseExtended[];
  folders: FolderResponse[];
}
